File: src/main/java/core/procedures/math/complex/Angle.java
package core.procedures.math.complex;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.BigComplex;
import core.utils.Utils;

public final class Angle extends AFn {

  public Angle() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Number.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "angle";
  }

  @Override
  public Number apply1(Object arg) {
    return angle((Number) arg);
  }

  private Number angle(Number number) {
    if (Utils.INSTANCE.isZero(number)) {
      throw new ArithmeticException(getName() + ": undefined for 0");
    }
    return BigComplex.of(number).angle();
  }
}

File: src/main/java/core/procedures/math/complex/Angle.kt
package core.procedures.math.complex

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Complex
import core.utils.Utils

class Angle : AFn<Number?, Number>(name = "angle", isPure = true, arity = Exactly(1),
                                   mandatoryArgsTypes = arrayOf(Number::class.java)) {

    override operator fun invoke(arg: Number?) = angle(arg!!)

    private fun angle(number: Number) = when {
        Utils.isZero(number) -> throw ArithmeticException("$name: undefined for 0")
        else -> Complex.valueOf(number).angle()
    }
}
